homeassistant/components/sensor/sabnzbd.py

```diff
@@ -4,216 +4,72 @@
 
-    def __init__(self, sensor_type, sabnzbd_api, client_name):
+    def __init__(self, hass, sensor_type, sabnzbd_api_data, client_name):
         """Initialize the sensor."""
+        self.hass = hass
+
+        self._client_name = client_name
+        self._field_name = SENSOR_TYPES[sensor_type][2]
         self._name = SENSOR_TYPES[sensor_type][0]
-        self.sabnzbd_api = sabnzbd_api
-        self.type = sensor_type
-        self.client_name = client_name
+        self._sabnzbd_api = sabnzbd_api_data
         self._state = None
+        self._type = sensor_type
         self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
 
+        async_dispatcher_connect(hass, SIGNAL_SABNZBD_UPDATED,
+                                 self.update_state)
+
     @property
     def name(self):
         """Return the name of the sensor."""
-        return '{} {}'.format(self.client_name, self._name)
+        return '{} {}'.format(self._client_name, self._name)
 
     @property
     def state(self):
         """Return the state of the sensor."""
         return self._state
 
+    def should_poll(self):
+        """Don't poll. Will be updated by dispatcher signal."""
+        return False
+
     @property
     def unit_of_measurement(self):
         """Return the unit of measurement of this entity, if any."""
         return self._unit_of_measurement
 
-    @asyncio.coroutine
-    def async_refresh_sabnzbd_data(self):
-        """Call the throttled SABnzbd refresh method."""
-        from pysabnzbd import SabnzbdApiException
-        try:
-            yield from async_update_queue(self.sabnzbd_api)
-        except SabnzbdApiException:
-            _LOGGER.exception("Connection to SABnzbd API failed")
-
-    @asyncio.coroutine
-    def async_update(self):
+    def update_state(self, args):
         """Get the latest data and updates the states."""
-        yield from self.async_refresh_sabnzbd_data()
-
-        if self.sabnzbd_api.queue:
-            if self.type == 'current_status':
-                self._state = self.sabnzbd_api.queue.get('status')
-            elif self.type == 'speed':
-                mb_spd = float(self.sabnzbd_api.queue.get('kbpersec')) / 1024
-                self._state = round(mb_spd, 1)
-            elif self.type == 'queue_size':
-                self._state = self.sabnzbd_api.queue.get('mb')
-            elif self.type == 'queue_remaining':
-                self._state = self.sabnzbd_api.queue.get('mbleft')
-            elif self.type == 'disk_size':
-                self._state = self.sabnzbd_api.queue.get('diskspacetotal1')
-            elif self.type == 'disk_free':
-                self._state = self.sabnzbd_api.queue.get('diskspace1')
-            elif self.type == 'queue_count':
-                self._state = self.sabnzbd_api.queue.get('noofslots_total')
-            else:
-                self._state = 'Unknown'
+        self._state = self._sabnzbd_api.get_queue_field(self._field_name)
```

Comment:
Is this doing I/O ?

Reply:
No I/O, just a dict lookup

Reply:
Use `None` as default value to represent unknown state.